<a href="https://colab.research.google.com/github/deepikadhiman5517/linearRegression/blob/main/linearregression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**1 Write a function to generate an m+1 dimensional data set, of size n, consisting of m continuous independent
 variables (X) and one dependent variable (Y) defined as <p>
yi = xiβ + e <p>**
where,<p>
• e is a Gaussuan distribution with mean 0 and standard deviation (σ), representing the unexplained
variation in Y<p>
• β is a random vector of dimensionality m + 1, representing the coefficients of the linear relationship
between X and Y, and<p>
• ∀i ∈ [1, n], xi0 = 1<p>
The function should take the following parameters:<p>
• σ: The spread of noise in the output variable<p>
• n: The size of the data set<p>
• m: The number of indepedent variables<p>
Output from the function should be:<p>
• X: An n × m numpy array of independent variable values (with a 1 in the first column)<p>
• Y : The n × 1 numpy array of output values<p>
• β: The random coefficients used to generatre Y from X<p>


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
def fun1(n,m,s):#genrate values
    X = np.random.rand(n, m+1)
    X[:, 0] = 1
    e = np.random.normal(0, s, (n,1)) # where 0 is loc,s is for scaling, (n,1) is the size
    b = np.random.rand(m+1, 1)
    y = np.matmul(X, b) + e
    return X,y,b


**2 Write a function that learns the parameters of a linear regression line given inputs**<p>
• X: An n × m numpy array of independent variable values<p>
• Y : The n × 1 numpy array of output values<p>
• k: the number of iteractions (epochs)<p>
• τ : the threshold on change in Cost function value from the previous to current iteration<p>
The function should implement the Gradient Descent algorithm as discussed in class that initialises β with
random values and then updates these values in each iteraction by moving in the the direction defined by
the partial derivative of the cost function with respect to each of the coefficients. The function should use
only one loop that ends after a number of iterations (k) or a threshold on the change in cost function value
(τ ).
The output should be an m + 1 dimensional vector of coefficients and the final cost function value.


In [ ]:
def regression(x, y, epochs, threshold):
    l = 1 #learning rate
    n = x.shape[0]
    m = x.shape[1]
    weight = np.random.rand(m,1)*l
    pre_cost = float('inf') #initial cost to infinity
    for i in range(epochs):
        pred_y = np.matmul(x, weight)
        cost = (1/(n)) * np.sum((y - pred_y)**2)
        if i == 0:
          print("Cost at epoch - 1: ", cost)
        if abs(pre_cost - cost) <= threshold:
            print("threshold reached at epoch: ", i)
            break
        pre_cost = cost  #updating the cost value

        weight = weight - l *(-2/n) * np.matmul(X.T, (y - pred_y)) # updating weights
    return pre_cost, weight


In [ ]:
X, y, b= fun1(600, 5, 1)


final_cost,weight = regression(X, y, 200, 0.0002)
print("Final cost: ", final_cost)

Cost at epoch - 1:  2.48099831004479
Final cost:  5.155869744921164e+222


3 Create a report investigating how differen values of n and σ impact the ability for your linear regression
function to learn the coefficients, β, used to generate the output vector Y .

<table border="4px">
<tr>
<th>n(size)</th>
<th>s(scale)</th>
<th>cost</th>

</tr>
<tr>
<td>500</td>
<td>0.2</td>
<td>6.13</td>
</tr>
<tr>
<td>500</td>
<td>0.5</td>
<td>9.7</td>
</tr>
<tr>
<td>500</td>
<td>1</td>
<td>2.8</td>
</tr>
<tr>
<td>500</td>
<td>2</td>
<td>1.7</td>
</tr>

<tr>
<td>400</td>
<td>0.1</td>
<td>2.5</td>
</tr>

<tr>
<td>600</td>
<td>0.1</td>
<td>2.5</td>
</tr>

<tr>
<td>600</td>
<td>1</td>
<td>5.15</td>
</tr>
</table>